File: peca-android/app/src/main/java/org/peercast/core/LogViewerFragment.java
package org.peercast.core;

import android.content.Context;
import android.content.res.Configuration;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.ColorRes;
import android.support.v4.app.ListFragment;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.text.SpannableStringBuilder;
import android.text.method.LinkMovementMethod;
import android.text.style.URLSpan;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVParser;
import org.apache.commons.csv.CSVRecord;
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.math.NumberUtils;

import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * @author (c) 2015, T Yoshizawa
 * Dual licensed under the MIT or GPL licenses.
 */
public class LogViewerFragment extends ListFragment {
    private static final String TAG = "LogViewerFragment";
    private LogAdapter mAdapter;
    private boolean mNarrowWidth;
    private AsyncTask mParserTask;


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.logviewer_fragment, container, false);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setHasOptionsMenu(true);

        AppCompatActivity act = (AppCompatActivity) getActivity();
        ActionBar actionBar = act.getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setTitle(R.string.t_view_log);

        mNarrowWidth = getResources().getConfiguration().screenWidthDp < 800;

        mAdapter = new LogAdapter();
        setListAdapter(mAdapter);

        doLogParse();
    }

    private void doLogParse() {
        if (mParserTask != null)
            return;
        PeerCastApplication app = (PeerCastApplication)getActivity().getApplication();
        Collection<File> logFiles = app.getLogFiles();
        mParserTask = new LogParseTask().execute(logFiles.toArray(new File[0]));
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        mNarrowWidth = newConfig.screenWidthDp < 800;
        doLogParse();
    }

    private class LogParseTask extends AsyncTask<File, Void, List<CSVRecord>> {
        @Override
        protected void onPreExecute() {
            mAdapter.mRecords.clear();
        }

        @Override
        protected List<CSVRecord> doInBackground(File... files) {
            List<CSVRecord> records = new ArrayList<>(512);
            Charset utf8 = Charset.forName("utf8");
            for (File f : files) {
                try {
                    CSVParser parser = CSVParser.parse(f, utf8, CSVFormat.DEFAULT);
                    records.addAll(parser.getRecords());
                } catch (IOException e) {
                    Log.e(TAG, "LogParseTask", e);
                }
            }
            return records;
        }

        @Override
        protected void onPostExecute(List<CSVRecord> records) {
            //if (!mAdapter.isEmpty())
            //    mListView.setSelection(mAdapter.getCount() - 1);]
            mAdapter.mRecords = records;
            mAdapter.notifyDataSetChanged();
            mParserTask = null;
        }
    }

    private class LogAdapter extends BaseAdapter {
        private List<CSVRecord> mRecords = Collections.emptyList();

        @Override
        public int getCount() {
            return mRecords.size();
        }

        @Override
        public CSVRecord getItem(int position) {
            //return mRecords.get(position);
            return mRecords.get(getCount() - position - 1);
        }

        @Override
        public long getItemId(int position) {
            return 0;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            Context c = parent.getContext();
            if (convertView == null) {
                LayoutInflater inflater = LayoutInflater.from(c);
                convertView = inflater.inflate(R.layout.logviewer_line, parent, false);
                convertView.setTag(new Holder(convertView));
            }
            bindView(c, (Holder) convertView.getTag(), getItem(position));
            return convertView;
        }



        private void bindView(Context c, Holder holder, CSVRecord record) {
            holder.vTime.setText(record.get(0));
            String className = record.get(1);
            boolean isAppClasses = className.startsWith("org.peercast.");

            if (mNarrowWidth || className.length() > 64)
                className = StringUtils.substringAfterLast(className, ".");

            holder.vSource.setText(className + "#" + record.get(2) + "()");

            int level = NumberUtils.toInt(record.get(3), 0);
            @ColorRes int color = R.color.md_grey_800;
            if (isAppClasses)
                color = R.color.md_green_800;
            if (level >= 1000)//ERROR
                color = R.color.md_red_800;
            else if (level >= 900)//WARN
                color = R.color.md_orange_800;

            //holder.vMessage.setText(record.get(4));
            holder.setMessage(record.get(4));
            holder.vMessage.setTextColor(c.getResources().getColor(color));
        }


    }

    private static class Holder {
        final TextView vTime;
        final TextView vMessage;
        final TextView vSource;

        Holder(View v) {
            vTime = (TextView) v.findViewById(R.id.vTime);
            vMessage = (TextView) v.findViewById(R.id.vMessage);
            vMessage.setMovementMethod(LinkMovementMethod.getInstance());
            vSource = (TextView) v.findViewById(R.id.vSource);
        }

        private static final Pattern URL_PATTERN = Pattern.compile("http://[^,\\s]+");

        void setMessage(String s){
            SpannableStringBuilder ssb = new SpannableStringBuilder(s);
            Matcher ma = URL_PATTERN.matcher(s);
            int start = 0;
            while (ma.find(start)) {
                start = ma.start();
                int end = ma.end();
                ssb.setSpan(new URLSpan(s.substring(start, end)), start, end, 0);
                start = ma.end();
            }
            vMessage.setText(ssb);
        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.logviewer_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_reload:
                doLogParse();
                return true;

            default:
                return super.onOptionsItemSelected(item);

        }
    }
}

File: app/src/main/java/org/peercast/core/LogViewerFragment.kt
package org.peercast.core

import android.content.Context
import android.os.Bundle
import android.text.SpannableStringBuilder
import android.text.method.LinkMovementMethod
import android.text.style.URLSpan
import android.view.*
import androidx.annotation.ColorInt
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import org.peercast.core.databinding.LogviewerLineBinding
import org.peercast.pecaport.PecaPort
import kotlin.coroutines.CoroutineContext

/**
 * @author (c) 2014-2019, T Yoshizawa
 * @licenses Dual licensed under the MIT or GPL licenses.
 */
class LogViewerFragment : Fragment(), CoroutineScope, PeerCastActivity.NestedScrollFragment {
    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main + job

    private lateinit var job: Job
    private val adapter = LogAdapter()

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        doLogParse()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        setHasOptionsMenu(true)
        job = Job()
        val c = inflater.context
        return RecyclerView(c).also {
            it.adapter = adapter
            it.layoutManager = LinearLayoutManager(c)
            it.addItemDecoration(DividerItemDecoration(c, DividerItemDecoration.VERTICAL))
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        job.cancel()
    }

    data class LogRecord(
            val time: CharSequence,
            val level: CharSequence,
            val source: CharSequence,
            val message: CharSequence
    ) {
        @ColorInt
        fun getMessageColor(c: Context): Int {
            return when {
                level == "ERROR" -> ContextCompat.getColor(c, R.color.md_red_800)
                level == "WARN" -> ContextCompat.getColor(c, R.color.md_orange_800)
                source.startsWith("org.peercast") -> ContextCompat.getColor(c, R.color.md_green_800)
                else -> ContextCompat.getColor(c, R.color.md_grey_800)
            }
        }
    }

    private fun doLogParse() = launch {
        val logFile = PecaPort.getLogFile(requireContext())
        if (!logFile.isFile)
            return@launch

        val s = logFile.reader().readText()
        adapter.records = RE_LOG_LINE.findAll(s).map { r ->
            r.groupValues.let {
                LogRecord(it[1], it[3], it[4], decorateMessage(it[5]))
            }
        }.toList()
    }

    private class ViewHolder(private val binding: LogviewerLineBinding)
        : RecyclerView.ViewHolder(binding.root) {

        init {
            binding.vMessage.movementMethod = LinkMovementMethod.getInstance()
        }

        fun bind(r: LogRecord){
            binding.r = r
            binding.executePendingBindings()
        }
    }

    private class LogAdapter : RecyclerView.Adapter<ViewHolder>() {
        var records = emptyList<LogRecord>()
            set(value) {
                field = value
                notifyDataSetChanged()
            }

        override fun getItemCount(): Int  = records.size

        fun getItem(position: Int): LogRecord {
            //return records[position]
            return records[itemCount - position - 1]
        }

        override fun getItemId(position: Int): Long = 0

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val inflater = LayoutInflater.from(parent.context)
            return ViewHolder(
                    LogviewerLineBinding.inflate(inflater, parent, false)
            )
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            holder.bind(getItem(position))
        }
    }

    private fun decorateMessage(s: String): CharSequence {
        return SpannableStringBuilder(s).also { ssb ->
            RE_URL.findAll(s).forEach { r ->
                val start = r.range.start
                val end = r.range.endInclusive + 1
                ssb.setSpan(URLSpan(s.substring(start, end)), start, end, 0)
            }
        }
    }


    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.logviewer_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_reload -> {
                doLogParse()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    companion object {
        private val RE_LOG_LINE = """^([\d:.]+)(.*?)\s+([A-Z]+)\s+(\S+)\s+- (.+?)$""".toRegex(
                setOf(RegexOption.DOT_MATCHES_ALL, RegexOption.MULTILINE)
        )
        private val RE_URL = """https?://[^,\s]+""".toRegex()
    }
}
